File: ssh_sessions/favorites.py
"""Numbered picker favorites, scoped to one catalog on one device."""
from dataclasses import dataclass
import hashlib
import json

from .catalog import CatalogError, atomic_write, identifier, locked

LOCAL = '@local'


def slot_number(value):
    if isinstance(value, bool) or str(value) not in '123456789' or len(str(value)) != 1:
        raise CatalogError('Choose a favorite slot from 1 to 9.')
    return str(value)


def target_id(value):
    if value == LOCAL:
        return LOCAL
    return identifier(value)


def validate_slots(slots):
    if not isinstance(slots, dict) or len(slots) > 9:
        raise CatalogError('Favorites must map slots 1–9 to machine IDs or @local.')
    result = {slot_number(slot): target_id(target) for slot, target in slots.items()}
    if len(set(result.values())) != len(result):
        raise CatalogError('Each machine or local terminal can occupy only one favorite slot.')
    return result


def target_name(target, machines):
    if target == LOCAL:
        return 'Local terminal'
    return next((m.name for m in machines if m.id == target), 'Missing machine: ' + target)


@dataclass(frozen=True)
class SavedFavorites:
    slots: dict[str, str]
    revision: str


class Favorites:
    def __init__(self, catalog):
        self.catalog = catalog
        self.path = catalog.device_path.with_suffix('.favorites.json')

    def load(self):
        raw = self.path.read_bytes() if self.path.exists() else b''
        slots = {}
        if self.path.exists():
            try:
                data = json.loads(raw) if len(raw) <= 8192 else None
                if not isinstance(data, dict) or set(data) != {'version', 'slots'} or type(data['version']) is not int or data['version'] != 1:
                    raise ValueError()
                slots = validate_slots(data['slots'])
            except (ValueError, TypeError, KeyError) as error:
                raise CatalogError('Local favorites are invalid; keep a backup and repair the .favorites.json file.') from error
        return SavedFavorites(slots, hashlib.sha256(raw).hexdigest())

    def _write(self, slots):
        raw = (json.dumps({'version': 1, 'slots': validate_slots(slots)}, indent=2) + '\n').encode()
        atomic_write(self.path, raw)

    def assign(self, slot, target, expected=None):
        slot = slot_number(slot)
        if target is not None:
            target = target_id(target)
        with locked(self.catalog.lock_path):
            saved = self.load()
            if expected is not None and saved.revision != expected:
                raise CatalogError('Favorites changed in another tab. Reopen Favorites before saving.')
            if target not in (None, LOCAL) and not any(m.id == target for m in self.catalog.load().machines):
                raise CatalogError('That machine is no longer in the catalog. Reload before assigning a favorite.')
            slots = {key: value for key, value in saved.slots.items() if key != slot and value != target}
            if target is not None:
                slots[slot] = target
            self._write(slots)

    def seed(self, slots):
        """Install portable defaults once; preserve later user edits, including empty slots."""
        slots = validate_slots(slots)
        with locked(self.catalog.lock_path):
            if self.path.exists():
                self.load()
                return False
            machine_ids = {m.id for m in self.catalog.load().machines}
            if set(slots.values()) - machine_ids - {LOCAL}:
                raise CatalogError('Favorite defaults reference a machine missing from the catalog.')
            self._write(slots)
            return True

File: src/favorites.rs
use crate::catalog::{Catalog, digest, identifier, locked, write_json};
use anyhow::{Context, Result, ensure};
use serde::{Deserialize, Serialize};
use std::{
    collections::{BTreeMap, HashSet},
    fs,
};

pub const LOCAL: &str = "@local";
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Favorites {
    pub version: u8,
    pub slots: BTreeMap<String, String>,
}
impl Default for Favorites {
    fn default() -> Self {
        Self {
            version: 1,
            slots: BTreeMap::new(),
        }
    }
}
impl Favorites {
    pub fn load(catalog: &Catalog) -> Result<(Self, String)> {
        let path = catalog.device_path.with_extension("favorites.json");
        if !path.exists() {
            return Ok((Self::default(), digest(b"")));
        }
        let raw = fs::read(path)?;
        ensure!(raw.len() <= 8192, "Favorites file is too large.");
        let value: Self = serde_json::from_slice(&raw)
            .context("Invalid favorites file. Keep a backup before repairing it.")?;
        value.validate()?;
        Ok((value, digest(&raw)))
    }
    fn validate(&self) -> Result<()> {
        ensure!(self.version == 1, "Unsupported favorites version.");
        let mut targets = HashSet::new();
        for (slot, target) in &self.slots {
            validate_slot(slot)?;
            if target != LOCAL {
                identifier(target)?;
            }
            ensure!(
                targets.insert(target),
                "A target cannot occupy multiple favorite slots."
            );
        }
        Ok(())
    }
    pub fn assign(
        catalog: &Catalog,
        slot: &str,
        target: Option<&str>,
        expected: Option<&str>,
    ) -> Result<()> {
        validate_slot(slot)?;
        let _guard = locked(&catalog.lock_path)?;
        let (mut value, revision) = Self::load(catalog)?;
        if let Some(expected) = expected {
            ensure!(
                revision == expected,
                "Favorites changed in another tab. Reload before saving."
            );
        }
        if let Some(target) = target {
            ensure!(
                target == LOCAL || catalog.load()?.machines.iter().any(|m| m.id == target),
                "Select a machine from the current catalog."
            );
            value.slots.retain(|_, t| t != target);
            value.slots.insert(slot.into(), target.into());
        } else {
            value.slots.remove(slot);
        }
        value.validate()?;
        write_json(
            &catalog.device_path.with_extension("favorites.json"),
            &value,
        )
    }
}
pub fn validate_slot(slot: &str) -> Result<()> {
    ensure!(
        slot.len() == 1 && (b'1'..=b'9').contains(&slot.as_bytes()[0]),
        "Choose a favorite number from 1 to 9."
    );
    Ok(())
}
